File: parser.py
"""LLM 返回解析 + 容错处理

最多重试 3 次，支持：
- JSON 格式（直接解析）
- markdown code block 包裹的 JSON
- 纯文本中提取数字
"""
from __future__ import annotations
import json
import re
import logging
from typing import Any, Dict, Optional, List

logger = logging.getLogger("ai-werewolf")


def parse_json_response(text: str) -> Optional[Dict[str, Any]]:
    """从 LLM 响应中提取 JSON dict，支持多种格式容错"""
    if not text:
        return None

    # 尝试1: 直接解析整个响应
    result = _try_parse_json(text)
    if result and isinstance(result, dict):
        return result

    # 尝试2: 提取 ```json ... ``` code block
    result = _extract_code_block(text)
    if result and isinstance(result, dict):
        return result

    # 尝试3: 找最外层的 { ... }
    result = _extract_braces(text)
    if result and isinstance(result, dict):
        return result

    # 尝试4: 修复常见错误（单引号、缺少引号等）
    result = _try_fix_and_parse(text)
    if result and isinstance(result, dict):
        return result

    logger.warning(f"无法解析 JSON 响应: {text[:200]}")
    return None


def parse_action(text: str, field: str,
                 valid_ids: List[int], default: Optional[int] = None) -> Optional[int]:
    """从响应中提取某个字段对应的玩家ID

    Args:
        text: LLM 原始响应
        field: 字段名 (kill, check, vote, shoot, poison)
        valid_ids: 合法的玩家ID列表
        default: 解析失败时的默认值
    """
    parsed = parse_json_response(text)
    if parsed:
        val = parsed.get(field)
        if val is not None:
            int_val = _to_int(val)
            if int_val is not None and int_val in valid_ids:
                return int_val

    # 兜底: 从文本中直接提取数字
    int_val = _extract_number_from_text(text, valid_ids)
    if int_val is not None:
        return int_val

    logger.warning(f"无法从响应中提取有效 {field}: {text[:200]}")
    return default


def parse_speech(text: str) -> str:
    """提取发言内容"""
    parsed = parse_json_response(text)
    if parsed and "speech" in parsed:
        return str(parsed["speech"])

    # 如果不是 JSON 格式，直接返回原文（去掉可能的 markdown 标记）
    cleaned = text.strip()
    if cleaned.startswith("```"):
        lines = cleaned.split("\n")
        cleaned = "\n".join(lines[1:-1] if lines[-1].strip() == "```" else lines[1:])
    return cleaned


def parse_witch_action(text: str) -> Dict[str, Any]:
    """解析女巫行动"""
    parsed = parse_json_response(text)
    if parsed:
        return {
            "save": bool(parsed.get("save", False)),
            "poison": _to_int(parsed.get("poison")),
        }

    # 容错: 从文本推断
    save = False
    poison = None
    if "救" in text or "使用解药" in text:
        save = True
    if "毒" in text:
        nums = re.findall(r'\d+', text)
        if nums:
            poison = int(nums[0])
    return {"save": save, "poison": poison}


def parse_bool_action(text: str, field: str) -> bool:
    """解析布尔型行动"""
    parsed = parse_json_response(text)
    if parsed:
        val = parsed.get(field)
        if val is not None:
            if isinstance(val, bool):
                return val
            if isinstance(val, str):
                return val.lower() in ("true", "yes", "是", "1")
            return bool(val)
    return False


# ─── 内部辅助 ───

def _try_parse_json(text: str) -> Optional[Any]:
    try:
        return json.loads(text.strip())
    except (json.JSONDecodeError, ValueError):
        return None


def _extract_code_block(text: str) -> Optional[Any]:
    """提取 ```json ... ``` 或 ``` ... ``` 中的内容"""
    patterns = [
        r'```json\s*\n?(.*?)\n?\s*```',
        r'```\s*\n?(.*?)\n?\s*```',
    ]
    for pat in patterns:
        m = re.search(pat, text, re.DOTALL)
        if m:
            return _try_parse_json(m.group(1))
    return None


def _extract_braces(text: str) -> Optional[Any]:
    """提取最外层 { ... }"""
    depth = 0
    start = -1
    for i, ch in enumerate(text):
        if ch == '{':
            if depth == 0:
                start = i
            depth += 1
        elif ch == '}':
            depth -= 1
            if depth == 0 and start >= 0:
                candidate = text[start:i + 1]
                result = _try_parse_json(candidate)
                if result is not None:
                    return result
                # 尝试修复
                result = _try_fix_and_parse(candidate)
                if result is not None:
                    return result
    return None


def _try_fix_and_parse(text: str) -> Optional[Any]:
    """修复常见 JSON 错误"""
    # 替换单引号为双引号
    fixed = text.replace("'", '"')
    # 修复没有引号的 key
    fixed = re.sub(r'(\{|,)\s*(\w+)\s*:', r'\1"\2":', fixed)
    # 修复 trailing comma
    fixed = re.sub(r',\s*}', '}', fixed)
    fixed = re.sub(r',\s*\]', ']', fixed)
    # 修复 Python bool
    fixed = fixed.replace("True", "true").replace("False", "false")
    fixed = fixed.replace("None", "null")
    try:
        return json.loads(fixed)
    except (json.JSONDecodeError, ValueError):
        return None


def _to_int(val: Any) -> Optional[int]:
    """将各种格式的值转为 int"""
    if val is None:
        return None
    if isinstance(val, int):
        return val
    if isinstance(val, float) and val.is_integer():
        return int(val)
    if isinstance(val, str):
        m = re.search(r'\d+', val)
        if m:
            return int(m.group())
    return None


def _extract_number_from_text(text: str, valid_ids: List[int]) -> Optional[int]:
    """从纯文本中提取合法的玩家 ID"""
    nums = re.findall(r'\b(\d+)\b', text)
    for n_str in reversed(nums):  # 取最后一个数字（通常是最终答案）
        n = int(n_str)
        if n in valid_ids:
            return n
    return None

File: lib/parser.ts
/**
 * LLM 响应解析 + 容错 —— 对应原 Python 版 parser.py
 *
 * 最多支持 4 层容错：
 * 1. 直接 JSON.parse
 * 2. 提取 ```json ... ``` code block
 * 3. 提取最外层 { ... }
 * 4. 修复常见错误后重试
 * 5. 从纯文本中提取数字兜底
 */

export interface ParsedObject {
  [key: string]: unknown;
}

export function parseJsonResponse(text: string): ParsedObject | null {
  if (!text) return null;

  // 尝试1: 直接解析
  const direct = tryParseJson(text);
  if (direct && typeof direct === 'object' && !Array.isArray(direct)) {
    return direct as ParsedObject;
  }

  // 尝试2: 提取 code block
  const fromBlock = extractCodeBlock(text);
  if (fromBlock && typeof fromBlock === 'object' && !Array.isArray(fromBlock)) {
    return fromBlock as ParsedObject;
  }

  // 尝试3: 提取最外层 { ... }
  const fromBraces = extractBraces(text);
  if (fromBraces && typeof fromBraces === 'object' && !Array.isArray(fromBraces)) {
    return fromBraces as ParsedObject;
  }

  // 尝试4: 修复后解析
  const fixed = tryFixAndParse(text);
  if (fixed && typeof fixed === 'object' && !Array.isArray(fixed)) {
    return fixed as ParsedObject;
  }

  return null;
}

export function parseAction(
  text: string,
  field: string,
  validIds: number[],
  defaultValue: number | null = null,
): number | null {
  const parsed = parseJsonResponse(text);
  if (parsed) {
    const val = parsed[field];
    if (val !== undefined && val !== null) {
      const intVal = toInt(val);
      if (intVal !== null && validIds.includes(intVal)) {
        return intVal;
      }
    }
  }

  // 兜底: 从文本中提取合法 ID
  const num = extractNumberFromText(text, validIds);
  if (num !== null) return num;

  return defaultValue;
}

export function parseSpeech(text: string): string {
  const parsed = parseJsonResponse(text);
  if (parsed && typeof parsed.speech === 'string') {
    return parsed.speech;
  }

  // 非 JSON 时直接返回原文，去掉 markdown 标记
  let cleaned = text.trim();
  if (cleaned.startsWith('```')) {
    const lines = cleaned.split('\n');
    const isClosed = lines[lines.length - 1].trim() === '```';
    cleaned = lines.slice(1, isClosed ? -1 : undefined).join('\n');
  }
  return cleaned;
}

export interface WitchAction {
  save: boolean;
  poison: number | null;
}

export function parseWitchAction(text: string): WitchAction {
  const parsed = parseJsonResponse(text);
  if (parsed) {
    return {
      save: Boolean(parsed.save ?? false),
      poison: toInt(parsed.poison ?? null),
    };
  }

  // 容错: 从文本推断
  let save = false;
  let poison: number | null = null;
  if (text.includes('救') || text.includes('使用解药')) {
    save = true;
  }
  if (text.includes('毒')) {
    const nums = text.match(/\d+/g);
    if (nums && nums.length) {
      poison = parseInt(nums[0], 10);
    }
  }
  return { save, poison };
}

export function parseBoolAction(text: string, field: string): boolean {
  const parsed = parseJsonResponse(text);
  if (parsed) {
    const val = parsed[field];
    if (val !== undefined && val !== null) {
      if (typeof val === 'boolean') return val;
      if (typeof val === 'string') {
        return ['true', 'yes', '是', '1'].includes(val.toLowerCase());
      }
      return Boolean(val);
    }
  }
  return false;
}

// ─── 内部辅助 ──────────────────────────────────────────────────

function tryParseJson(text: string): unknown {
  try {
    return JSON.parse(text.trim());
  } catch {
    return null;
  }
}

function extractCodeBlock(text: string): unknown {
  const patterns = [
    /```json\s*\n?([\s\S]*?)\n?\s*```/,
    /```\s*\n?([\s\S]*?)\n?\s*```/,
  ];
  for (const pat of patterns) {
    const m = text.match(pat);
    if (m) {
      const result = tryParseJson(m[1]);
      if (result !== null) return result;
      const fixed = tryFixAndParse(m[1]);
      if (fixed !== null) return fixed;
    }
  }
  return null;
}

function extractBraces(text: string): unknown {
  let depth = 0;
  let start = -1;
  for (let i = 0; i < text.length; i++) {
    const ch = text[i];
    if (ch === '{') {
      if (depth === 0) start = i;
      depth += 1;
    } else if (ch === '}') {
      depth -= 1;
      if (depth === 0 && start >= 0) {
        const candidate = text.slice(start, i + 1);
        const result = tryParseJson(candidate);
        if (result !== null) return result;
        const fixed = tryFixAndParse(candidate);
        if (fixed !== null) return fixed;
      }
    }
  }
  return null;
}

function tryFixAndParse(text: string): unknown {
  let fixed = text.replace(/'/g, '"');
  // 给无引号的 key 加引号
  fixed = fixed.replace(/(\{|,)\s*(\w+)\s*:/g, '$1"$2":');
  // 修复 trailing comma
  fixed = fixed.replace(/,\s*}/g, '}');
  fixed = fixed.replace(/,\s*]/g, ']');
  // 修复 Python bool
  fixed = fixed.replace(/\bTrue\b/g, 'true').replace(/\bFalse\b/g, 'false');
  fixed = fixed.replace(/\bNone\b/g, 'null');
  try {
    return JSON.parse(fixed);
  } catch {
    return null;
  }
}

function toInt(val: unknown): number | null {
  if (val === null || val === undefined) return null;
  if (typeof val === 'number') {
    return Number.isInteger(val) ? val : null;
  }
  if (typeof val === 'string') {
    const m = val.match(/\d+/);
    if (m) return parseInt(m[0], 10);
  }
  return null;
}

function extractNumberFromText(text: string, validIds: number[]): number | null {
  const matches = text.match(/\b(\d+)\b/g);
  if (!matches) return null;
  // 取最后一个匹配的合法 ID（通常是最终答案）
  for (let i = matches.length - 1; i >= 0; i--) {
    const n = parseInt(matches[i], 10);
    if (validIds.includes(n)) return n;
  }
  return null;
}
